lib/autoload/course/assessment/programming_test_case_report.rb

```diff
@@ -143,6 +161,44 @@ def error_type
       @test_case.search('error')[0]['type']
     end
 
+    # If there's a failure, return the failure type and failure message.
+    #
+    # @return [String] A combined string with the failure type and failure message
+    # @return [nil] If there's no failure
+    def failure_message
+      return nil unless failed?
+      "#{failure_type}: #{_failure_message}"
+    end
+
+    # If there's a failure, return the contents of the failure tag.
+    # This contains the full traceback.
+    #
+    # @return [String] Full traceback of failure.
+    # @return [nil] If there's no failure
+    def failure_contents
+      return nil unless failed?
+      @test_case.search('failure')[0].children.to_s
+    end
+
+    # If there's a failure, return the failure message attribute.
+    # For internal use only
+    #
+    # @return [String] The message attribute
+    # @return [nil] If there's no failure
+    def _failure_message
```

Comment:
make this private, I noticed that there's a `_error_message`,  better to change that too. and how about to rename it to `failure_contents` ?